File: src/ros381_base/src/odometry.cpp
#include "nav_msgs/msg/odometry.hpp"
#include "../include/signal.hpp"
#include "rclcpp/rclcpp.hpp"
#include "ros381_interfaces/msg/float3.hpp"
#include "ros381_interfaces/srv/update_pose.hpp"
#include "tf2/LinearMath/Quaternion.h"

class OdometryNode : public rclcpp::Node
{
  public:
    OdometryNode() : Node("odometry")
    {
        this->declare_parameter("L", 0.297);
        L_ = this->get_parameter("L").as_double();
        L_recip_ = 1 / L_;
        this->set_parameter(rclcpp::Parameter("use_sim_time", true));

        passive_vel_sub_ = this->create_subscription<ros381_interfaces::msg::Float3>(
            "base_encoders", 10, std::bind(&OdometryNode::callback_passive_vel, this, std::placeholders::_1));

        odom_pub_ = this->create_publisher<nav_msgs::msg::Odometry>("odom", 10);

        update_srv_ = this->create_service<ros381_interfaces::srv::UpdatePose>(
            "update_pose",
            std::bind(&OdometryNode::callback_update_pose, this, std::placeholders::_1, std::placeholders::_2));

        RCLCPP_INFO(this->get_logger(), "Odometry node is running.");
    }

  private:
    void callback_update_pose(const std::shared_ptr<ros381_interfaces::srv::UpdatePose::Request> request,
                              std::shared_ptr<ros381_interfaces::srv::UpdatePose::Response> response)
    {

        response->success = false;

        bool update_x = (request->type / 100) % 10;
        bool update_y = (request->type / 10) % 10;
        bool update_phi = (request->type / 1) % 10;

        if (update_x)
            x_base_ = request->x;
        if (update_y)
            y_base_ = request->y;
        if (update_phi)
            phi_base_ = request->phi;

        RCLCPP_INFO(this->get_logger(), "New pose :\nx = %.2f mm\ny = %.2f mm\nphi = %.2f rad", x_base_, y_base_,
                    phi_base_);

        response->success = update_x || update_y || update_phi;
    }

    void callback_passive_vel(const ros381_interfaces::msg::Float3::SharedPtr msg)
    {
        v_right_ = msg->float3[0];
        v_left_ = msg->float3[1];
        dt_ = msg->float3[2];
        calculate_odometry();
    }

    void calculate_odometry()
    {
        if (odom_initialized_)
        {
            v_base_ = (v_right_ + v_left_) * 0.5;
            w_base_ = (v_right_ - v_left_) * L_recip_;
            mid_angle_ = phi_base_ + w_base_ * dt_ * 0.5;

            x_base_ += v_base_ * cos(mid_angle_) * dt_;
            y_base_ += v_base_ * sin(mid_angle_) * dt_;
            phi_base_ += w_base_ * dt_;
            wrap_ptr(&phi_base_, -M_PI, M_PI);

            // RCLCPP_INFO (this->get_logger (),
            //              "\nv = %.3f\nw = %.3f\nx = %.3f\ny = %.3f\nphi = %.3f",
            //              v_base_, w_base_, x_base_, y_base_, phi_base_);
            this->publish_odometry();
        }
        else
        {
            odom_initialized_ = true;
            RCLCPP_INFO(this->get_logger(), "Odometry initialized!");
        }
    }

    void publish_odometry()
    {
        auto msg = nav_msgs::msg::Odometry();
        // Set the header
        msg.header.stamp = now();
        msg.header.frame_id = "odom";     // Adjust as needed
        msg.child_frame_id = "base_link"; // Adjust as needed

        // Set position
        msg.pose.pose.position.x = x_base_;
        msg.pose.pose.position.y = y_base_;
        msg.pose.pose.position.z = 0.0;

        // Convert yaw (phi_base_) to quaternion
        tf2::Quaternion q;
        q.setRPY(0, 0, phi_base_);
        msg.pose.pose.orientation.x = q.x();
        msg.pose.pose.orientation.y = q.y();
        msg.pose.pose.orientation.z = q.z();
        msg.pose.pose.orientation.w = q.w();

        // Set linear and angular velocity
        msg.twist.twist.linear.x = v_base_;
        msg.twist.twist.linear.y = 0.0;
        msg.twist.twist.linear.z = 0.0;

        msg.twist.twist.angular.x = 0.0;
        msg.twist.twist.angular.y = 0.0;
        msg.twist.twist.angular.z = w_base_;
        odom_pub_->publish(msg);
    }

    double L_;            // [m]
    double L_recip_;      // [1/m]
    double v_right_;      // [m/s]
    double v_left_;       // [m/s]
    double dt_;           // [s]
    double v_base_;       // [m/s]
    double w_base_;       // [rad/s]
    double phi_base_ = 0; // [rad]
    double mid_angle_;    // [rad]
    double x_base_ = 0;   // [m]
    double y_base_ = 0;   // [m]
    bool odom_initialized_ = false;

    rclcpp::Subscription<ros381_interfaces::msg::Float3>::SharedPtr passive_vel_sub_;
    rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_pub_;
    rclcpp::Service<ros381_interfaces::srv::UpdatePose>::SharedPtr update_srv_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<OdometryNode>();
    sleep(1);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/ros381_webots/ros381_webots/odometry.py
import rclpy
from rclpy.node import Node
from ros381_interfaces.msg import Float3
from ros381_interfaces.srv import UpdatePose
from nav_msgs.msg import Odometry
from geometry_msgs.msg import Quaternion
import math

R_RIGHT = 0.072 * 0.5
R_LEFT = 0.072 * 0.5
L = 0.297
L_recip = 1 / L


class OdometryW:
    def init(self, webots_node, properties):
        self.robot_ = webots_node.robot
        self.robot_name = self.robot_.getName()

        try:
            rclpy.init(args=None)
        except:
            pass

        self.node_ = rclpy.create_node("odometry", namespace=f"/{self.robot_name}")
        self.encoders_sub_ = self.node_.create_subscription(
            Float3,
            f"/{self.robot_name}/base_encoders",
            self.encoders_callback,
            10,
        )
        self.odometry_pub_ = self.node_.create_publisher(
            Odometry, f"/{self.robot_name}/odom", 10
        )
        self.node_.get_logger().info(
            f"Webots odometry node for {self.robot_name} is initialized."
        )
        self.odometry_msg_ = Odometry()
        self.odometry_msg_.pose.pose.position.x = 0.0
        self.odometry_msg_.pose.pose.position.y = 0.0
        self.odometry_msg_.pose.pose.position.z = 0.0
        self.odometry_msg_.pose.pose.orientation.x = 0.0
        self.odometry_msg_.pose.pose.orientation.y = 0.0
        self.odometry_msg_.pose.pose.orientation.z = 0.0
        self.odometry_msg_.pose.pose.orientation.w = 0.0
        self.odometry_msg_.twist.twist.linear.x = 0.0
        self.odometry_msg_.twist.twist.linear.y = 0.0
        self.odometry_msg_.twist.twist.linear.z = 0.0
        self.odometry_msg_.twist.twist.angular.x = 0.0
        self.odometry_msg_.twist.twist.angular.y = 0.0
        self.odometry_msg_.twist.twist.angular.z = 0.0
        self.odometry_msg_.header.stamp = self.node_.get_clock().now().to_msg()

        self.x_base_ = 0.0
        self.y_base_ = 0.0
        self.phi_base_ = 0.0
        
        self.update_srv_ = self.node_.create_service(
            UpdatePose,
            'update_pose',
            self.callback_update_pose
        )
        self.odom_init_ = False;
    
    
    def callback_update_pose(self, request, response):
        response.success = False

        update_x = (request.type // 100) % 10
        update_y = (request.type // 10) % 10
        update_phi = (request.type // 1) % 10

        if update_x:
            self.x_base_ = request.x
        if update_y:
            self.y_base_ = request.y
        if update_phi:
            self.phi_base_ = request.phi

        self.node_.get_logger().info(
            f"New pose:\nx = {self.x_base_:.2f} mm\ny = {self.y_base_:.2f} mm\nphi = {self.phi_base_:.2f} rad"
        )

        response.success = bool(update_x or update_y or update_phi)
        return response

    def encoders_callback(self, msg):
        if msg.float3[2] > 0.0 and not math.isnan(msg.float3[0]) and not math.isnan(msg.float3[1]) and not math.isnan(msg.float3[2]):
            self.calculate_odometry(msg.float3[0], msg.float3[1], msg.float3[2])

    def calculate_odometry(self, v_right, v_left, dt):
        if self.odom_init_:
            # 1. izracunaj odometriju
            self.odometry_msg_.header.stamp = self.node_.get_clock().now().to_msg()
            v_base = (v_right + v_left) * 0.5;
            w_base = (v_right - v_left) * L_recip;
            
            self.node_.get_logger().info(
                f"w = {w_base:.2f}"
            )
            mid_angle = self.phi_base_ + w_base * dt * 0.5;
            
            self.x_base_ += v_base * math.cos(mid_angle) * dt;
            self.y_base_ += v_base * math.sin(mid_angle) * dt;
            self.phi_base_ += w_base * dt;
            diff = 2 * math.pi
            while self.phi_base_ > math.pi:
                self.phi_base_ -= diff
            while self.phi_base_ < -math.pi:
                self.phi_base_ += diff
            
            # 2. pub odometriju
            q = Quaternion()
            cy = math.cos(self.phi_base_ * 0.5)
            sy = math.sin(self.phi_base_ * 0.5)
            self.odometry_msg_.pose.pose.position.x = self.x_base_
            self.odometry_msg_.pose.pose.position.y = self.y_base_
            self.odometry_msg_.pose.pose.orientation.z = sy
            self.odometry_msg_.pose.pose.orientation.w = cy
            self.odometry_msg_.twist.twist.linear.x = v_base
            self.odometry_msg_.twist.twist.angular.z = w_base
            self.odometry_pub_.publish(self.odometry_msg_)
        else:
            self.odom_init_ = True

    def step(self):
        rclpy.spin_once(self.node_, timeout_sec=0)
